Add lingo-rep-raycast extension

asashay opened the pull request:
## Description

The LingoRep Raycast extension makes it easy to translate text between languages. Anyone can use it for quick translations. Once you sign in, you can save your translations and use the spaced repetition algorithm in the web app to help memorize them effectively. More about it on the web site - https://lingorep.com/

## Screencast
[Screencast video](https://www.loom.com/share/06dff3205bb846018ba3a43a8e4a5813?sid=b9f52c85-ebf3-4f1f-bc00-d39b5625e303)

<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented on extensions/lingo-rep-raycast-extension/package.json:
```diff
@@ -0,0 +1,58 @@
+    {
+      "name": "index",
+      "title": "Lingorep Translate",
+      "subtitle": "Translate, Save and Repeat Translations",
```
```suggestion
```
Should not be used for description

pernielsentikaer reviewed (changes requested):
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- Could you make a screencast of the flow so I can see how it works 😊

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

asashay commented:
@pernielsentikaer thank you for getting back to me! I've added screencast and description - please check them out

asashay commented on extensions/lingo-rep-raycast-extension/package.json:
```diff
@@ -0,0 +1,57 @@
+    "dev-raycast": "",
+    "fix-lint": "ray lint --fix",
+    "lint": "ray lint",
+    "publish": "npx @raycast/api@latest publish"
```
@pernielsentikaer sure, done

asashay replied:
@pernielsentikaer it's done, could you take a look?

asashay replied:
@pernielsentikaer changed, please take a look

asashay commented:
hey @pernielsentikaer , I also shortened my repo URL from `lingo-rep-raycast-extension.git` to `lingo-rep-raycast.git`. I think it makes more sense from the readability stand point.

Will it be ok to publish it like that? Or this it better to resubmit it?

pernielsentikaer commented:
oAuth is not meant to be used like that with multiple services, did you see ![Raycast 2025-02-11 at 08 26 48](https://github.com/user-attachments/assets/8ae5c673-f427-4e10-9121-b2b8d5f57c9d)?

When I login I'm try to login it just ends in this view

![Raycast 2025-02-11 at 08 26 48](https://github.com/user-attachments/assets/d640b078-4e7b-44ba-9bc5-4eb7d84ad174)

Is there a reason why you need to login, it should more be a login to LingoRep.com?

asashay commented:
hey @pernielsentikaer, thanks for getting back to me.

`oAuth is not meant to be used like that with multiple services, did you see [this page](https://developers.raycast.com/api-reference/oauth)?`
yes, I checked it when I implemented oAuth 

I've made some changes - now after login user is redirected to the main view of the Raycast, please take a look.

As for the necessity of login - in our web app we use Google and Github to authenticate users. Thus I've implemented them here as well. Without being authenticated users won't be able to save translations to the database.

The idea is as following:
- users translate and save words on their desktop using this Raycast extension
- later they can review and repeat them in order to memorize them. This is done in the web extension

Cheers

pernielsentikaer commented:
It feels a bit weird from a user perspective getting the actions when you're first entering the extension - I'm not sure if we can do anything to make this a better experience

asashay commented:
hey @pernielsentikaer , could you elaborate more on `feels a bit weird from a user perspective getting the actions when you're first entering the extension`? I don't think I understand why it is an issue.

Currently, initial flow for user is:
1. Install extension
2. Configure language pairs
3. Translate between them

Logging in - is optional step for user, if they want to save and repeat words later.

asashay commented:
hey @pernielsentikaer , thanks for details and for QAing it thoroughly!

It was a bug with authorization in my code. I just fixed it and pushed a new version.

Could you take a look?

pernielsentikaer commented:
There is some hardcoded paths that can't be resolved and some security vulnerbilities
```
❯ npm run dev

> lingo-rep-raycast@0.1.0 dev
> npm run npm:install && npm run copy:shared-packages && npm-run-all --parallel watch:shared-packages dev:raycast


> lingo-rep-raycast@0.1.0 npm:install
> npm install --prefix /Users/alex/Work/Projects/alex/lingo-rep-monorepo/apps/lingo-rep-raycast

npm error code EACCES
npm error syscall mkdir
npm error path /Users/alex
npm error errno -13
npm error Error: EACCES: permission denied, mkdir '/Users/alex'
npm error     at async mkdir (node:internal/fs/promises:857:10)
npm error     at async Arborist.reify (/Users/pernielsentikaer/.nvm/versions/node/v22.12.0/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:126:7)
npm error     at async Install.exec (/Users/pernielsentikaer/.nvm/versions/node/v22.12.0/lib/node_modules/npm/lib/commands/install.js:150:5)
npm error     at async Npm.exec (/Users/pernielsentikaer/.nvm/versions/node/v22.12.0/lib/node_modules/npm/lib/npm.js:207:9)
npm error     at async module.exports (/Users/pernielsentikaer/.nvm/versions/node/v22.12.0/lib/node_modules/npm/lib/cli/entry.js:74:5) {
npm error   errno: -13,
npm error   code: 'EACCES',
npm error   syscall: 'mkdir',
npm error   path: '/Users/alex'
npm error }
npm error
npm error The operation was rejected by your operating system.
npm error It is likely you do not have the permissions to access this file as the current user
npm error
npm error If you believe this might be a permissions issue, please double-check the
npm error permissions of the file and its containing directories, or try running
npm error the command again as root/Administrator.
npm error A complete log of this run can be found in: /Users/pernielsentikaer/.npm/_logs/2025-02-23T07_31_36_529Z-debug-0.log
❯ npm audit
# npm audit report

axios  1.3.2 - 1.7.3
Severity: high
Server-Side Request Forgery in axios - https://github.com/advisories/GHSA-8hc4-vh64-cxmj
fix available via `npm audit fix --force`
Will install axios@1.7.9, which is outside the stated dependency range
node_modules/axios

esbuild  <=0.24.2
Severity: moderate
esbuild enables any website to send any requests to the development server and read the response - https://github.com/advisories/GHSA-67mh-4wv8-2f99
fix available via `npm audit fix`
node_modules/esbuild
  @raycast/api  >=1.86.0
  Depends on vulnerable versions of esbuild
  node_modules/@raycast/api

3 vulnerabilities (2 moderate, 1 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues, run:
  npm audit fix --force
 ```

asashay commented:
@pernielsentikaer thanks for checking it! 

I've fixed security vulnerability in axios. Another 2 are in the `@raycast/api`. I've updated it to latest version, but they are still there. Is there anything else I can do?

Also fixed package.json - now you should be able to run `npm run dev` without issues. 

Please take a look!

pernielsentikaer commented:
Where does this action save to?

![lingo-rep-raycast 2025-02-25 at 12 18 16](https://github.com/user-attachments/assets/32709fd9-7fbb-4a58-90f3-59aeda6985e5)

pernielsentikaer commented:
I found it, that's okay for now 🙂

Can we make sure that all actions in the action panel has icons?

asashay commented:
sure, added icons to all actions in the extension.

Is it ok now? @pernielsentikaer

pernielsentikaer commented:
@greptileai can you have a last look?

pernielsentikaer commented on extensions/lingo-rep-raycast/src/index.tsx:
```diff
@@ -0,0 +1,369 @@
+        },
+        jwt,
+      );
+      showToast({ title: "Translation saved" });
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+      showHUD("Translation saved", { clearRootSearch: true, popToRootType: PopToRootType.Immediate });
```
I would just pop to root directly here 😊

asashay replied:
I've removed it altogether, seems redundant with the toast notification above

asashay commented on extensions/lingo-rep-raycast/src/index.tsx:
```diff
@@ -0,0 +1,369 @@
+        </ActionPanel>
+      }
+    >
+      <List.Section title="Results" subtitle={`${_.get(translatedText, "translation.sentences[0].orig", "")}`}>
+        {translatedText?.translation?.dict
+          ? _.map(translatedText?.translation?.dict, (pos) => (
+              <SearchListItem key={pos.pos} searchResult={pos} actions={itemActions} />
+            ))
+          : translatedText?.translation?.sentences
+            ? _.map(translatedText?.translation.sentences, (sentence) => (
+                <List.Item title={sentence.trans ?? ""} key={sentence?.trans} actions={itemActions} />
+              ))
+            : null}
+      </List.Section>
```
there is no isLoading prop on the `Section` component

asashay commented:
@pernielsentikaer addressed the comments, please take a look

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
